---
title: "2017-07-22_AnalysisManiAgign"
author: "Thorsten Brach"
date: "22/7/2017"
output: html_document
---


# Deciding on trimLeft and silva or rdp

- Because I got so few taxa assigned down to species level, I checked whether there might be primer sequences left, simply by running Dada_Pipel also with trimLeft c(30,30). 
- I did assigned taxonomy to seqtab.nochim both with silva (version 128) and rdp (version 16). I got very few sequence variants (SVs) assigned down to species level in both cases
- I decided to stick with the combination where I got most SVs assigned to species levels

```{r, message = FALSE, warning = FALSE}
# the version trimLeft = c(10,10)
datapath <- "/Users/jvb740/MarieCurie_Work/MouseProject/ResultsAndProtocols/ManiAging_Results/16S_Sequencing/2017-07-13_DK_age_ManiAging/Dada2_Analysis/Dada_Taxonomy"
load(file.path(datapath, "rdp/Taxonomy_rdp16_minBoot50_allowMultT.RData"))

trim10_rdp <- taxa.species

load(file.path(datapath, "Silva_v128/Taxonomy_Silva128_minBoot50_allowMT.RData"))

trim10_silva <- taxa.species

datapath <- "/Users/jvb740/MarieCurie_Work/MouseProject/ResultsAndProtocols/ManiAging_Results/16S_Sequencing/2017-07-13_DK_age_ManiAging/Dada2_Analysis_trimLeft30/Dada_Taxonomy"
load(file.path(datapath, "rdp/Taxonomy_rdp16_minBoot50_allowMT.RData"))

trim30_rdp <- taxa.species

load(file.path(datapath, "Silva128/Taxonomy_Silva128_minBoot50_allowMT.RData"))

trim30_silva <- taxa.species

# First NB: with trim30 you got only 854 SVs while with trim 10 you got 869
SumNoNASpecies <- c(sum(!is.na(trim10_rdp[,"Species"])), sum(!is.na(trim10_silva[,"Species"])), sum(!is.na(trim30_rdp[,"Species"])), sum(!is.na(trim30_silva[,"Species"])))
TotalSV <- c(nrow(trim10_rdp), nrow(trim10_silva), nrow(trim30_rdp), nrow(trim30_silva))
DF <- data.frame(Type = c("trim10_rdp", "trim10_silva", "trim30_rdp", "trim30_silva"), SumNoNASpecies = SumNoNASpecies, TotalSV = TotalSV, Perc = 100*(SumNoNASpecies/TotalSV))

DF
```

- Silva with trimLeft = c(10,10) wins
- NB: I should compare how well the assignment from rdp overlaps with silva

# Load data and explore a bit

```{r, message = FALSE, warning = FALSE}
rm(list = ls())

# Load seqtab.nochim
datapath <- "/Users/jvb740/MarieCurie_Work/MouseProject/ResultsAndProtocols/ManiAging_Results/16S_Sequencing/2017-07-13_DK_age_ManiAging/Dada2_Analysis"
load(file.path(datapath, "Dada_Data/DenoisedData.RData"))

rm(ReadSummary, seqtab, mergers)

load(file.path(datapath, "Dada_Taxonomy/Silva_v128/Taxonomy_Silva128_minBoot50_allowMT.RData"))

rm(taxa, InputSave)


functionpath <- "/Users/jvb740/MarieCurie_Work/BackgroundKnowledge/16S_Learning/Dada_Pipel/Functions/"

source(file.path(functionpath, "Dada_TaxonomyFunctions.R"))
source(file.path(functionpath, "Dada_PlotFunctions.R"))

```

## The distribution of amplicons = SVs in the samples

- you will see you had 61 samples

```{r}
TrList <- AmpliconDistribution(seqtab.nochim, PCentage = 15)
TrList[[1]] # More than 150 of the SVs were only found in 1 sample
```

- For filtering on prevalence you might ask how many of the SVs are lost, and how many of the total reads linked to these SVs are lost
- with a prevalence of 15% you see that 

```{r}
PCremainingSV <- as.numeric(TrList[[5]][10,"CumPerCUnique"])
TrList[[2]] + geom_vline(xintercept = 10) + geom_hline(yintercept = PCremainingSV)
PCremainingSV
TrList[[4]] # More than 94% of reads would remain, prevalence of 15% accounts to be present in at least 10 samples
```

- so a 15% prevalence filtering would keep less than 50% of the SVs, but more than 94% of reads

## Determine for each taxonomic level the percentage of sequences that could not be attributed unambiguously with the used minBoot

```{r, message = FALSE, warning = FALSE}
NAPerC <- NAPerCForTaxLevel(taxa.species)

NAPerC


# look at the Genus-species with multiple species assignment (at the first 5)
head(unname(taxa.species[grep(pattern = "/", taxa.species[,'Species']), c('Genus', 'Species')]), 5)

# # # Number and percentage of assigned taxonomic levels related to the penetrance of the amplicons
# # TLvsPenetrance <- TaxLevelvsPenetrance(taxa = taxa.species, seqtab = seqtab.nochim)
# 
# # Taxonomic assignments compared to the abundance of the sequences
# TLvsAb <- TaxLevelvsAbundance(taxa = taxa.species, seqtab = seqtab.nochim, Level = "Genus")
# 
# TLvsAb[[2]]
# # could be done for other levels
```

# Analyses in Phyloseq

## Generate the phyloseq object


```{r, message = FALSE, warning = FALSE}
# generate the data frame with the sample data
sample.path <- "/Users/jvb740/MarieCurie_Work/MouseProject/ResultsAndProtocols/ManiAging_Results/Results_RAnalysis"
load(file.path(sample.path, "ManiAging.RData"))

# I start with the quality report from the sequencing
samdf <- ManiAging[["QR"]]
# the order in samdf has to fit to seqtab.nochim
samples.out <- rownames(seqtab.nochim)
Study <- sapply(strsplit(samples.out, "-"), `[`, 1)
IDfirst <- paste(sapply(strsplit(samples.out, "-"), `[`, 2))
IDsecond <- paste(sapply(strsplit(samples.out, "-"), `[`, 3))
IDorder <- vector(mode = "character", length = length(samples.out))
IDorder[IDsecond == "NA"] <- IDfirst[IDsecond == "NA"]
IDorder[IDsecond != "NA"] <- paste(IDfirst[IDsecond != "NA"], ".", IDsecond[IDsecond != "NA"], sep = "")

samdf <- samdf[match(IDorder, samdf$Mouse_ID),]
rownames(samdf) <- samples.out


# Generate the phyloseq file
ps <- phyloseq(otu_table(seqtab, taxa_are_rows=FALSE), tax_table(taxa.species), sample_data(samdf))
ps

# # some playing with phyloseq
# ntaxa(ps)
# nsamples(ps)
# sample_names(ps)
# sample_variables(ps)
# get_variable(ps, sample_variables(ps)[2])
# sample_data(ps)$Age
# rank_names(ps)
# colnames(tax_table(ps))
# get_taxa_unique(ps, "Phylum")
# sample_sums(ps)
# taxa_sums(ps) # abundance of the taxa over all samples (sum)
# get_taxa(ps, i = sample_names(ps)[1]) # so i = 1 refers to a sample
# get_sample(ps, taxa_names(ps)[1])

```





