---
title: "Taxonomic assignation"
author: "Camila"
date: "December 15, 2016"
output: html_document
---
# Required packages

```{r}
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")

```

# Links

- <http://benjjneb.github.io/dada2/assign.html#species-assignment>
- The phyloseq website
    - <https://joey711.github.io/phyloseq/>

# Questions

- how dependent are the results on the training data (i.e. Silva or other database)?
- can you improve the plots, phyloseq training
- what allowMultiple should we use

# Background

- minBoot: default = 50, is the minimum number out of 100 bootstraps that return the same genus (taxonomy) for a sequence so the sequence is assigned that taxonomy
- For human microbiome data and V4 sequencing, we typically see a bit under half of the abundant ribosomal sequence variants (RSVs) assigned unambiguously to species-level, and a bit under 2/3rds assigned ambiguously (ie. allowing multiple species assignment). This fraction drops off in less sampled environments and for rare RSVs.

# In case you restart, load your data

```{r}

datapath <- "/Users/jvb740/MarieCurie_Work/BackgroundKnowledge/16S_Learning/DanFunD_243_maxEE-1"
load(file.path(datapath, "Dada_Data/DenoisedData.RData"))

load(file.path(datapath, "TaxAssign/TaxonomicData.RData"))


```

# Analyse a bit your seqtab, Do you want to exclude rare sequences?

```{r}
FinalNumbers <- data.frame(Sample = rownames(seqtab), UniqueAmplicons = rowSums(seqtab != 0), NoAmplicons = rowSums(seqtab))
FinalNumbersSeq <- data.frame(Sequence = colnames(seqtab), InNumberSamples = colSums(seqtab != 0), TotalAmplicons = colSums(seqtab))
FinalNumbersSeq <- group_by(FinalNumbersSeq, InNumberSamples)
AmpliconDistribution <- dplyr::summarise(FinalNumbersSeq, UniqueAmplicons = n(), TotalAmplicons = sum(TotalAmplicons))
# compare, table could also provide you the unique amplicons
#AmpliconDistr <- as.data.frame(table(colSums(seqtab != 0)))


AmpliconDistribution$CumSumUnique <- rev(cumsum(rev(AmpliconDistribution$UniqueAmplicons)))
AmpliconDistribution$CumPerCUnique <- rev(cumsum(rev(AmpliconDistribution$UniqueAmplicons/ncol(seqtab))))
AmpliconDistribution$CumSumTotal <- rev(cumsum(rev(AmpliconDistribution$TotalAmplicons)))
AmpliconDistribution$CumPerCTotal <- rev(cumsum(rev(AmpliconDistribution$TotalAmplicons/sum(colSums(seqtab)))))


## Distribution plots

# The number of samples the unique amplicons are present in
Tr <- ggplot(AmpliconDistribution, aes(x = InNumberSamples, y = UniqueAmplicons))
Tr <- Tr + geom_point(col = "#E69F00", size = 3) +
        xlab("Present in Number of Samples") +
        ylab("Unique Amplicons") +
        theme_bw() +
        theme(panel.grid.minor = element_blank(),
              panel.grid.major.y = element_blank(),
              panel.grid.major.x = element_line(color = "#999999", size = .15))

# Cumulative Percentage of Unique Amplicons
Tr1 <- ggplot(AmpliconDistribution, aes(x = InNumberSamples, y = CumPerCUnique))
Tr1 <- Tr1 + geom_point(col = "#E69F00", size = 3) +
        xlab("Present in Number of Samples") +
        ylab("Cumulative Percentage of unique Amplicons") +
        theme_bw() +
        theme(panel.grid.minor = element_blank(),
              panel.grid.major.y = element_blank(),
              panel.grid.major.x = element_line(color = "#999999", size = .15))

# The number of samples the total amplicons are present in
Tr2 <- ggplot(AmpliconDistribution, aes(x = InNumberSamples, y = TotalAmplicons))
Tr2 <- Tr2 + geom_point(col = "#E69F00", size = 3) +
        xlab("Present in Number of Samples") +
        ylab("Total Amplicons") +
        theme_bw() +
        theme(panel.grid.minor = element_blank(),
              panel.grid.major.y = element_blank(),
              panel.grid.major.x = element_line(color = "#999999", size = .15))

# Cumulative Percentage of Total Amplicons
# probably the best plot to decide on the number of samples an amplicon is supposed to be present to be kept
TenPercentValue <- ceiling(0.1*dim(seqtab)[1])
Diff <- AmpliconDistribution$InNumberSamples - TenPercentValue
index <- which.max(Diff[Diff<0]) + which.min(Diff[Diff>=0])
PCKeptAtTenPCValue <- AmpliconDistribution$CumPerCTotal[index]
Tr3 <- ggplot(AmpliconDistribution, aes(x = InNumberSamples, y = CumPerCTotal))
Tr3 <- Tr3 + geom_point(col = "#E69F00", size = 3) +
        xlab("Present in Number of Samples") +
        ylab("Cumulative Percentage of total Amplicons") +
        geom_hline(yintercept = PCKeptAtTenPCValue, lty =  "dashed") +
        geom_vline(xintercept = TenPercentValue, lty = 'dashed') +
        ggtitle(paste("Total Amplicons", AmpliconDistribution$CumSumTotal[1], "; 10%:", index, "samples;", round(PCKeptAtTenPCValue, 3), "% of Total Amplicons remain")) +
        theme_bw() +
        theme(panel.grid.minor = element_blank(),
              panel.grid.major.y = element_line(color = "#999999", size = .15),
              panel.grid.major.x = element_line(color = "#999999", size = .15))



## In case you prefer a histogram
# FinalNumbers2 <- data.frame(InNoSamples = colSums(seqtab != 0))
# 
# if(nrow(seqtab) < 10) {binwidth = 1}
# if(nrow(seqtab) > 10 && length(SampleNames) < 100) {binwidth = 2}
# if(nrow(seqtab) > 100) {binwidth = 3}
#         
# Trr <- ggplot(data = FinalNumbers2, aes(x = InNoSamples))  + 
#         geom_histogram(binwidth = binwidth, col = "black", fill = "#E69F00")+
#         geom_rug() +
#         xlab("Present in No Samples") + 
#         ylab("Count") +
#         theme_bw() + 
#         ggtitle(paste("Total No of unique amplicons:", dim(seqtab)[2], "Only in 1 sample:", sum(FinalNumbers2 ==1))) +
#         theme(panel.grid.minor = element_blank(),
#               panel.grid.major.y = element_blank(),
#               panel.grid.major.x = element_line(color = "#999999", size = .15)) +
#         coord_cartesian(ylim = c(0,150))
##########


```

# Define the Paths to the reference fastq files

```{r}
PathToRefs <- "/Users/jvb740/MarieCurie_Work/BackgroundKnowledge/16S_Learning/AssignTaxonomy"
RefDataBase <- "silva_nr_v123_train_set.fa.gz"
RefDB <- file.path(PathToRefs, RefDataBase)

if(!file.exists(RefDB)){
        stop("could not find the reference database for assignTaxonomy")
}

SpeciesDB <- "silva_species_assignment_v123.fa.gz"
SpecDB <- file.path(PathToRefs, SpeciesDB)

if(!file.exists(SpecDB)){
        stop("could not find the reference database for addSpecies")
}
```

# NB: the next three steps can be jumbed over if taxa.80 and taxa.80.species and co have already been collected, therefore commented

# Use assignTaxonomy for taxonomic assignment down to genus for your seqtab

```{r}
# seqs<-c(colnames(seqtab))
## ATTENTION: takes a bit less than a minute per 100 sequences, so for 3600 sequences it takes just below 30 minutes
ptm <- proc.time()
taxa.80 <- assignTaxonomy(seqtab, refFasta = RefDB,verbose = TRUE, minBoot=80) 
proc.time() - ptm

ptm <- proc.time()
taxa.50 <- assignTaxonomy(seqtab, refFasta = RefDB,verbose = TRUE, minBoot=50) 
proc.time() - ptm

# NB: taxa.80 is a matrix: nrows is number of sequences, ncol is number of taxonomic levels (usually 6: Kingdom, Phylum, Class, Order, Family, Genus)
identical(colnames(seqtab), rownames(taxa.80)) # TRUE
identical(colnames(seqtab), rownames(taxa.50)) # TRUE
```


# Use addSpecies to add species level to the taxa matrixes

- The addSpecies command takes pretty long, so better on the server

```{r}
# ATTENTION: these takes long: 
ptm <- proc.time()
taxa.80.species <- addSpecies(taxa.80,refFasta = SpecDB,verbose = TRUE, allowMultiple = 3)
proc.time() - ptm


# ATTENTION: these takes long: 
ptm <- proc.time()
taxa.50.species <- addSpecies(taxa.50,refFasta = SpecDB,verbose = TRUE, allowMultiple = 3)
proc.time() - ptm


save(taxa.80, taxa.50, taxa.80.species, taxa.50.species, file = file.path(PathToRefs, "TaxonomicData.RData"))
```


# Exploratory Analysis of the Taxa tables

## Overview of taxa matrices 

```{r}
# check that addSpecies does not change the Kingdom to Genus levels
identical(taxa.80.species[,1:6], taxa.80) #TRUE

# look at the Genus-species with multiple species assignment
unname(taxa.80.species[grep(pattern = "/", taxa.80.species[,'Species']), c('Genus', 'Species')])
```

## Determine for each taxonomic level the percentage of sequences that could not be attributed unambiguously with the used minBoot

- you need to source Dada_TaxonomyFunctions.R for the following functions

```{r}
# Number and percentage of assigned taxonomic levels for all amplicons
NAPerC.80 <- NAPerCForTaxLevel(taxa.80.species)
NAPerC.50 <- NAPerCForTaxLevel(taxa.50.species)

# Number and percentage of assigned taxonomic levels related to the penetrance of the amplicons
TLvsPenetrance.80 <- TaxLevelvsPenetrance(taxa = taxa.80.species, seqtab = seqtab)

# Taxonomic assignments compared to the abundance of the sequences
TLvsAb.80 <- TaxLevelvsAbundance(taxa = taxa.80.species, seqtab = seqtab, Level = "Genus")

TLvsAb.80[[2]]


# Extra training
M1 <- mergers.nochim[[1]]
Uniques <- getUniques(M1)
# same as M1$abundance
identical(M1$abundance, unname(Uniques))
```


# Going into phyloseq 
```{r}
## Generate sample data
# sample_data in phyloseq is just a data.frame with rownames = Sample names where information on these samples is stored
# e.g
SampleNames <- rownames(seqtab)
Age <- sample(x = 20:70, size = length(SampleNames), replace = TRUE)
samdf <- data.frame(Sample = SampleNames, Age = Age)
samdf$Above50 <- Age > 50
rownames(samdf) <- SampleNames
rbinom(n = length(SampleNames), size = 1, prob = .5)

# Generate the phyloseq file
ps <- phyloseq(otu_table(seqtab, taxa_are_rows=FALSE), tax_table(taxa.80.species), sample_data(samdf))
ps
ntaxa(ps)
nsamples(ps)
sample_names(ps)
sample_variables(ps)
get_variable(ps, sample_variables(ps)[2])
sample_data(ps)$Age
rank_names(ps)
colnames(tax_table(ps))
get_taxa_unique(ps, "Phylum")
sample_sums(ps)
taxa_sums(ps) # abundance of the taxa over all samples (sum)
get_taxa(ps, i = sample_names(ps)[1]) # so i = 1 refers to a sample
get_sample(ps, taxa_names(ps)[1])


plot_richness(ps)
plot_richness(ps, measures = "Observed")
plot_richness(ps, measures = "Shannon")

Tr <- plot_richness(ps, measures = "Observed")
Tr <- Tr + 
        theme_bw() + 
        ylab("Richness") +
        geom_point(col = "#E69F00", size = 3)


# example of using only the taxa that are present in at least 2 samples
Above2Taxa <- taxa_names(ps)[colSums(otu_table(ps) != 0) > 1]
psTop <- transform_sample_counts(ps, function(OTU) OTU/sum(OTU)) # change to relative Abundances here
psTop = prune_taxa(Above2Taxa, psTop)
sample_sums(psTop) # NOTE if you do it this way around: the relative abundances do not sum to 1
plot_bar(psTop, x="Sample", fill="Family")
plot_bar(psTop, x="Sample", fill="Phylum")

psTop2 = prune_taxa(Above2Taxa, ps)
psTop2 <- transform_sample_counts(psTop2, function(OTU) OTU/sum(OTU)) # change to relative Abundances here
sample_sums(psTop2) # sum to 1
plot_bar(psTop2, x="Sample", fill="Family")
plot_bar(psTop2, x="Sample", fill="Phylum")





```



